src/libraries/System.Runtime.InteropServices.RuntimeInformation/src/System/Runtime/InteropServices/RuntimeInformation/RuntimeInformation.PlatformVersion.cs

```diff
@@ -0,0 +1,94 @@
+        }
+
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major)
+            => IsOSPlatform(osPlatform) && Environment.OSVersion.Version.Major >= major;
+
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor)
+            => IsOSPlatform(osPlatform) && IsOSVersionOrLater(major, minor, int.MinValue, int.MinValue);
+
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build)
+            => IsOSPlatform(osPlatform) && IsOSVersionOrLater(major, minor, build, int.MinValue);
+
+        public static bool IsOSPlatformOrLater(OSPlatform osPlatform, int major, int minor, int build, int revision)
+            => IsOSPlatform(osPlatform) && IsOSVersionOrLater(major, minor, build, revision);
+
+        public static bool IsOSPlatformEarlierThan(string platformName)
+        {
+            (OSPlatform platform, Version version) = Parse(platformName);
+
+            return IsOSPlatformEarlierThan(platform, version.Major, version.Minor, version.Build, version.Revision);
+        }
+
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major)
+            => IsOSPlatform(osPlatform) && Environment.OSVersion.Version.Major < major;
+
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major, int minor)
+            => IsOSPlatform(osPlatform) && !IsOSVersionOrLater(major, minor, int.MinValue, int.MinValue);
+
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major, int minor, int build)
+            => IsOSPlatform(osPlatform) && !IsOSVersionOrLater(major, minor, build, int.MinValue);
+
+        public static bool IsOSPlatformEarlierThan(OSPlatform osPlatform, int major, int minor, int build, int revision)
+            => IsOSPlatform(osPlatform) && !IsOSVersionOrLater(major, minor, build, revision);
+
+        private static bool IsOSVersionOrLater(int major, int minor, int build, int revision)
+        {
+            Version current = Environment.OSVersion.Version;
+            if (current.Major != major)
+            {
+                return current.Major > major;
+            }
+            if (current.Minor != minor)
+            {
+                return current.Minor > minor;
+            }
+            if (current.Build != build)
+            {
+                return current.Build > build;
+            }
+
+            return current.Revision >= revision;
+        }
+
+        private static (OSPlatform, Version) Parse(string platformName)
+        {
+            if (platformName == null)
+            {
+                throw new ArgumentNullException(nameof(platformName));
+            }
+            if (platformName.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_EmptyValue, nameof(platformName));
+            }
+
+            for (int i = 0; i < platformName.Length; i++)
```

Comment:
@terrajobst shared some comments with me on this:

> Ideally, we'd follow NuGet parsing rules: if there are no periods in the version number, then each digit is assumed to be separate. Thus "14" means "1.4".

As far as whether or not to permit digits in the middle of the name, he also recommended we match the behavior that MSBuild and NuGet have here.  Given `abc123def45` as a platform name, we should match how NuGet would parse that.  I'm going to try checking NuGet's behavior for that...

Reply:
Here's the documentation for NuGet's target framework name support:
https://docs.microsoft.com/en-us/nuget/reference/target-frameworks

Reply:
[Here's the implementation](https://github.com/NuGet/NuGet.Client/blob/c278cc47700700aa643efb821574d931dfdc425e/src/NuGet.Core/NuGet.Frameworks/NuGetFrameworkFactory.cs#L317). NuGet also starts from the beginning of the string, so it also does not support numbers within the platform/framework name.

Reply:
good point, fixed

Reply:
Sorry we crossed paths on this. Please revert the loop change per the new info on NuGet's parsing.  Sorry about that.  Instead of using Version.Parse though, we should use parsing that is consistent with NuGet's, where:

- If there are dots in the version number, then we can use Version.Parse
- If there are no dots in the version number, then we would treat each digit as a separate version segment

That would have "windows81" get parsed as "Windows 8.1" for example.  @terrajobst pointed out that would yield behavior consistent across MSBuild and NuGet where these monikers are also specified as strings, per the NuGet spec above.

Reply:
> If there are no dots in the version number, then we would treat each digit as a separate version segment

this could cause problems. What "windows10" would mean? windows 10.0 or 1.0? I added the xml docs and a clear exception message and imo we can enforce the mandatory major and minor versions separated with dot

Reply:
Yeah, even though it would be nice to align with NuGet here, I have to agree with you that we shouldn't do this.  We even called that out in the other comments where we decided "ios14" should not be supported because of that ambiguity.  If we inherit the "infer the dots" logic, then "ios14" would get interpreted as "ios 1.4" and that would not be leading people into the behavior they'd be expecting.  Requiring the dots removes all ambiguity and we then don't support the ambiguous case.  I'm sold.

Sorry for the back-and-forth.